### DISCUSSION TITLE
Web3 Unleashed Rental Marketplace issue-- ParserError: Expected '{' but got reserved keyword 'override'

### DISCUSSION THREAD

**Discussion by insanity54** [ConsenSys-archive/truffle#5715] [General Q&A] [answered]
Thank you for the wonderful web series, Web3 unleashed! I've been enjoying the videos.

I'm following the guide, [Web3 Unleashed: Write a Rentable NFT Smart Contract](https://trufflesuite.com/guides/rentable-nft/#web3-unleashed-write-a-rentable-nft-smart-contract).

I'm trying to implement a NFT rental marketplace, and I've encountered a problem while following the tutorial.

```
chris@zed-zed:~/Documents/minetest-web3_/truffle$ yarn; truffle migrate --network goerli
yarn install v1.22.19
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.30s.

Compiling your contracts...
===========================
> Compiling ./contracts/ethereum/ERC4907.sol
> Compiling ./contracts/ethereum/IERC4907.sol
> Compiling ./contracts/ethereum/RentableVips.sol
> Compiling ./contracts/ethereum/SimpleStorage.sol

CompileError: project:/contracts/ethereum/ERC4907.sol:23:82: ParserError: Expected '{' but got reserved keyword 'override'
  function setUser(uint256 tokenId, address user, uint64 expires) public virtual override {
                                                                                 ^------^
,project:/contracts/ethereum/RentableVips.sol:24:1: ParserError: Expected pragma, import directive or contract/interface/library definition.
Footer
^----^

Compilation failed. See above.
    at /home/chris/.config/yarn/global/node_modules/truffle/build/webpack:/packages/compile-solidity/dist/run.js:95:1
    at Generator.next (<anonymous>)
    at fulfilled (/home/chris/.config/yarn/global/node_modules/truffle/build/webpack:/packages/compile-solidity/dist/run.js:28:43)
Truffle v5.6.5 (core: 5.6.5)
Node v18.12.1
```

Not really sure what to do about this. I think I'll start from scratch and see if I missed anything in the tutorial, then report back.

EDIT: I found the [the source repo for the nft rental marketplace guide](https://github.com/trufflesuite/unleashed_nft_rental_marketplace) and cloned it. Now I can rule out any faulty copy-pasting on my end. I ran `yarn`, updated `truffle-config.js`, and inserted my own `.env` file. Then I ran the migrate command and here's what I'm seeing

```
chris@zed-zed:~/Documents/unleashed_nft_rental_marketplace/truffle$ truffle migrate --network goerli

Compiling your contracts...
===========================
> Compiling ./contracts/ERC4907.sololc-bin. Attempt #1
> Compiling ./contracts/IERC4907.sol
> Compiling ./contracts/Marketplace.sol
> Compiling ./contracts/RentableNft.sol
> Compiling @openzeppelin/contracts/interfaces/IERC165.sol
> Compiling @openzeppelin/contracts/interfaces/IERC721.sol
> Compiling @openzeppelin/contracts/security/ReentrancyGuard.sol
> Compiling @openzeppelin/contracts/token/ERC721/ERC721.sol
> Compiling @openzeppelin/contracts/token/ERC721/IERC721.sol
> Compiling @openzeppelin/contracts/token/ERC721/IERC721Receiver.sol
> Compiling @openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol
> Compiling @openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol
> Compiling @openzeppelin/contracts/utils/Address.sol
> Compiling @openzeppelin/contracts/utils/Context.sol
> Compiling @openzeppelin/contracts/utils/Counters.sol
> Compiling @openzeppelin/contracts/utils/Strings.sol
> Compiling @openzeppelin/contracts/utils/introspection/ERC165.sol
> Compiling @openzeppelin/contracts/utils/introspection/IERC165.sol
> Compiling @openzeppelin/contracts/utils/math/Math.sol
> Compiling @openzeppelin/contracts/utils/structs/EnumerableSet.sol

CompileError: TypeError: Function has override specified but does not override anything.
  --> project:/contracts/ERC4907.sol:86:22:
   |
86 |   ) internal virtual override {
   |                      ^^^^^^^^

,TypeError: Wrong argument count for function call: 3 arguments given but expected 4.
  --> project:/contracts/ERC4907.sol:87:5:
   |
87 |     super._beforeTokenTransfer(from, to, tokenId);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Compilation failed. See above.
    at /home/chris/.config/yarn/global/node_modules/truffle/build/webpack:/packages/compile-solidity/dist/run.js:95:1
    at Generator.next (<anonymous>)
    at fulfilled (/home/chris/.config/yarn/global/node_modules/truffle/build/webpack:/packages/compile-solidity/dist/run.js:28:43)
Truffle v5.6.5 (core: 5.6.5)
Node v18.12.1
```

Research time! I will look into these errors.

**EDIT2**: 

The first error is `TypeError: Function has override specified but does not override anything.
  --> project:/contracts/ERC4907.sol:86`

I looked up the source for the two classes(?) that our ERC4907 extends. That's ERC721URIStorage, and IERC4907.

_beforeTokenTransfer isn't part of ERC721URIStorage.sol, so it must be part of IERC4907.sol. Here is that file, with comments omitted to save space

```solidity
// IERC4907.sol
// SPDX-License-Identifier: MIT
pragma solidity >=0.4.22 <0.9.0;
interface IERC4907 {
  event UpdateUser(uint256 indexed tokenId, address indexed user, uint64 expires);
  function setUser(uint256 tokenId, address user, uint64 expires) external;
  function userOf(uint256 tokenId) external view returns(address);
  function userExpires(uint256 tokenId) external view returns(uint256);
}
```
There is no `_beforeTokenTransfer` here! I think this is the cause of the error. `_beforeTokenTransfer` actually exists in ERC721.sol, which is what ERC731URIStorage extends.

```solidity
// ERC731URIStorage.sol
abstract contract ERC721URIStorage is ERC721 {...}
```

The second error is `TypeError: Wrong argument count for function call: 3 arguments given but expected 4.  --> project:/contracts/ERC4907.sol:87`

I looked at ERC721 and saw how _beforeTokenTransfer is defined

```solidity
    function _beforeTokenTransfer(
        address from,
        address to,
        uint256, /* firstTokenId */
        uint256 batchSize
    ) internal virtual {
        if (batchSize > 1) {
            if (from != address(0)) {
                _balances[from] -= batchSize;
            }
            if (to != address(0)) {
                _balances[to] += batchSize;
            }
        }
    }
```
There are indeed 4 parameters. I don't understand why `firstTokenId` is commented out. maybe because it's not being referenced in the function? Commented to prevent linter error?

Well, I understand the errors now, but I'm not sure how to fix them.

**Comment by insanity54** [comment, marked as answer]
I think I figured it out. There was a [breaking change in _beforeTokenTransfer](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/7c5f6bc2c8743d83443fa46395d75f2f3f99054a/CHANGELOG.md#breaking-changes) about 9 days ago which requires _beforeTokenTransfer to have a fourth parameter.

Here's what my _beforeTokenTransfer function looks like now, and this was able to deploy without an error.

```solidity
 function _beforeTokenTransfer(
    address from,
    address to,
    uint256 tokenId,
    uint256 /** batch **/
  ) internal virtual override {
    super._beforeTokenTransfer(from, to, tokenId, 1);

    if (from != to && _users[tokenId].user != address(0)) {
      delete _users[tokenId];
      emit UpdateUser(tokenId, address(0), 0);
    }
  }
```

Looks like the guide uses openzeppelin 4.7.3, and yarn installed 4.8.0.


EDIT: Forgot to mention that I solved the `ParserError: Expected '{' but got reserved keyword 'override'` issue by changing compilers.solc.version setting in truffle-config.js to "0.8.17"

**Reply by cds-amal** [reply]
Nice sleuthing @insanity54 !

**Reply by lpurdy01** [reply]
@insanity54 Fantastic catch. I think you are still the only source on the internet with this solution.
